daemons/based/pacemaker-based.c

```diff
@@ -126,6 +126,21 @@ setup_stand_alone(GError **error)
     return pcmk_rc_ok;
 }
 
+static void
+based_metadata(void)
+{
+    const char *name = "pacemaker-based";
+    const char *desc_short = "Cluster Information Base manager options";
+    const char *desc_long = "Cluster options used by Pacemaker's Cluster "
+                            "Information Base manager";
+
+    gchar *s = pcmk__cluster_option_metadata(name, desc_short, desc_long,
+                                             pcmk__opt_context_based);
+
+    printf("%s", s);
```

Comment:
Yep, we already have formatted output of XML (such as `crm_resource --show-metadata`) via `out->output_xml()`. Text output shows it as-is, XML output wraps it in a CDATA block.

Since we're defining the XML in this case (rather than passing on something from an agent), we don't need to use CDATA. In fact, in this case, text output shouldn't show XML at all. Text output can be a simple list, and XML output OCF-like.

If it winds up being trivial to use formatted output here after creating the functions, we might as well, but I wouldn't waste time specifically for it.

Reply:
> Yep, we already have formatted output of XML (such as crm_resource --show-metadata) via out->output_xml(). Text output shows it as-is, XML output wraps it in a CDATA block.

It's too late to change it, but I feel like it would've been better to transform it to a simple list for text output and show the XML as-is for XML output. In other words, my first inclination would've been to model this on `pcs resource describe`.

> Since we're defining the XML in this case (rather than passing on something from an agent), we don't need to use CDATA. In fact, in this case, text output shouldn't show XML at all. Text output can be a simple list, and XML output OCF-like.

Yep that was my plan. As mentioned above, it could be pcs-like. Although pcs may want to get XML output and parse it themselves, for reliability and ultimate control of formatting.

> If it winds up being trivial ...

I suspect it will (create a text object, use once and destroy). Could be wrong.

Reply:
> I feel like it would've been better to transform it to a simple list for text output

I suppose the counter-argument is that the default output should match that of `/path/to/agent meta-data`.

Reply:
> > Yep, we already have formatted output of XML (such as crm_resource --show-metadata) via out->output_xml(). Text output shows it as-is, XML output wraps it in a CDATA block.
> 
> It's too late to change it, but I feel like it would've been better to transform it to a simple list for text output and show the XML as-is for XML output. In other words, my first inclination would've been to model this on `pcs resource describe`.

It would be cool to have readable text output for agent metadata (like "Parameters: ...."). But the XML output will always need a CDATA block because we can't define an API schema for externally generated XML. Then again someone might want the raw XML metadata -- I suppose we could keep that as the default text output and finally make use of the `--text-fancy` option to show it in a readable way.
 
> > Since we're defining the XML in this case (rather than passing on something from an agent), we don't need to use CDATA. In fact, in this case, text output shouldn't show XML at all. Text output can be a simple list, and XML output OCF-like.
> 
> Yep that was my plan. As mentioned above, it could be pcs-like. Although pcs may want to get XML output and parse it themselves, for reliability and ultimate control of formatting.

Definitely, the whole point of XML output is that anyone parsing the output should use that.

Reply:
> But the XML output will always need a CDATA block because we can't define an API schema for externally generated XML.

OCF agents' metadata is *supposed* to be able to validate against the OCF RA schema (even though we don't provide it currently), right? I mean, tools like pcs implicitly rely on that already, and so would a conversion to text output.

It's a loose dependence, in that we can skip some things in the XML that are absent from the schema and vice-versa. And that might still be enough of a problem to preclude us from ever outputting the raw XML -- some XML in a third-party file that would fail to validate, would cause the Pacemaker output to fail to validate.

Though I'm not 100% sure that "Pacemaker output fails to validate when it's given input (the agent) that fails to validate" should be a deal-breaker.

As I finish typing this, I remember that fence agent metadata is another problem. No longdescs. Extra `<getopt>` elements. Etc. You know what? Nevermind.

> I suppose we could keep that as the default text output and finally make use of the --text-fancy option to show it in a readable way.

https://projects.clusterlabs.org/T750